catalyst-toolbox-cli/src/cli/recovery/tally/mockchain.rs

```diff
@@ -0,0 +1,108 @@
+use chain_impl_mockchain::block::HeaderId;
+use chain_impl_mockchain::fragment::Fragment;
+use chain_impl_mockchain::ledger::Ledger;
+use jormungandr_lib::interfaces::PersistentFragmentLog;
+
+use crate::cli::recovery::tally::Error;
+use std::collections::vec_deque::VecDeque;
+
+pub fn generate_ledger_from_fragments<'block0>(
+    block0_header_id: HeaderId,
+    block0_fragments: impl Iterator<Item = &'block0 Fragment>,
+    logged_fragments: impl Iterator<Item = Result<PersistentFragmentLog, Error>>,
+) -> Result<(Ledger, VecDeque<PersistentFragmentLog>), Error> {
+    let ledger = Ledger::new(block0_header_id, block0_fragments).map_err(Error::LedgerError)?;
+    let parameters = ledger.get_ledger_parameters();
+    let block_date = ledger.date();
+
+    let mut fragments: VecDeque<PersistentFragmentLog> = VecDeque::new();
+    let mut tally_fragments: Vec<PersistentFragmentLog> = Vec::new();
+
+    // process fragments lazily
+    for fragment in logged_fragments {
+        match fragment {
+            Err(e) => {
+                println!("Error processing fragment: {:?}", e);
+            }
+            Ok(fragment_log) => {
+                if matches!(&fragment_log.fragment, Fragment::VoteTally(_)) {
+                    tally_fragments.push(fragment_log);
+                    continue;
+                }
+                match ledger.apply_fragment(&parameters, &fragment_log.fragment, block_date) {
```

Comment:
`apply_fragment` returns the new ledger, we need to re-assign it in case of success

Reply:
oops, sorry missed those

Reply:
I haven't pushed them still though 😅